pkg/pgmodel/pgx.go

```diff
@@ -1021,6 +1025,84 @@ func (q *pgxQuerier) LabelValues(labelName string) ([]string, error) {
 	return labelValues, nil
 }
 
+const GetLabelsSQL = "SELECT (key_value_array($1::int[])).*"
+
+type labelQuerier interface {
+	getLabelsForIds(ids []int64) (lls labels.Labels, err error)
+}
+
+func (q *pgxQuerier) getPrompbLabelsForIds(ids []int64) (lls []prompb.Label, err error) {
+	ll, err := q.getLabelsForIds(ids)
+	if err != nil {
+		return
+	}
+	lls = make([]prompb.Label, len(ll))
+	for i := range ll {
+		lls[i] = prompb.Label{Name: ll[i].Name, Value: ll[i].Value}
+	}
+	return
+}
+
+func (q *pgxQuerier) getLabelsForIds(ids []int64) (lls labels.Labels, err error) {
+	keys := make([]interface{}, len(ids))
+	values := make([]interface{}, len(ids))
+	for i := range ids {
+		keys[i] = ids[i]
+	}
+	numHits := q.labels.GetValues(keys, values)
+
+	if numHits < len(ids) {
+		err = q.fetchMissingLabels(keys[numHits:], ids[numHits:], values[numHits:])
+		if err != nil {
+			return
+		}
+	}
+
+	lls = make([]labels.Label, 0, len(values))
+	for i := range values {
+		lls = append(lls, values[i].(labels.Label))
+	}
+
+	return
+}
+
+func (q *pgxQuerier) fetchMissingLabels(misses []interface{}, missedIds []int64, newLabels []interface{}) error {
+	for i := range misses {
+		missedIds[i] = misses[i].(int64)
+	}
+	rows, err := q.conn.Query(context.Background(), GetLabelsSQL, missedIds)
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
```

Comment:
should make sure we get at most one row?

Reply:
should we use QueryRow instead?

Reply:
our shim doesn't expose that function ☹️